### PR TITLE
introduce morphological reconstruction operator `mreconstruct`

### PR DESCRIPTION
**Pull request by johnnychen94**
This is the first step to getting #60 into the codebase

TODO:

- [x] add benchmark
- [x] add alias `underbuild` and `overbuild`
- [x] add examples and detailed explanation on https://juliaimages.org/ImageMorphology.jl/dev/operators/
~- [ ] tweak the performance by splitting the loops to border part and inner part (via `EdgeIterator` from TiledInteration.jl)~
- [ ] absorb the optimized Bool version in #60

~Because LoopVectorization needs special handling and careful test coverage, I do not plan to utilize LV in this PR. We can try it in future PR.~

`hmaxima`/`hminima` and `regional_maxima`/`regional_minima` will be introduced in future PRs.

~@ThomasRetornaz do you think it necessary to introduce the term `underbuild`/`overbuild`?~ never mind, I'll just add them as alias.

cc: @ThomasRetornaz

### CODE REVIEW

**Comment by johnnychen94** [comment]
After some experiments, I realize that 1) splitting the forward/backward scan loop into the inner region and edge region, and 2) applying LoopVectorization probably won't work because this algorithm requires a fixed sequential order to properly propagate the values.

Without changing the algorithm, optimizations that I can think of is

- pad the marker image with sentinel values so that we can skip bounds check and save some computation (This boosts about 15% of the performance). Given that the loop order must be fixed, SIMD and other loop reorder tricks (e.g., `@turbo`) can hardly be applied here.
- short-circuit the loop if we know in prior that it's already the maximal/minimal value. (#60 does this for Bool value)
- specialized implementation for particular SE shapes. For instance, for 2D diamond shape SE, we can manually inline the `upper_se` and `lower_se` to avoid the array `getindex`. But I'm not sure how much performance we can get.

**Comment by johnnychen94** [comment]
I plan to merge this PR anyway and leave the optimization in future PR.

File: docs/operators/operators/mreconstruct.jl [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+# The default structuring element it uses is the box-shaped SE. Like the other operators,
+# you can pass `dims` keyword or positional argument `se` to `mreconstruct`. For instance,
+
+mreconstruct(dilate, img .- 0.3, img; dims=1) # reconstruction along colomn
+mreconstruct(dilate, img .- 0.3, img, strel_diamond(img)) # diamond shape SE
+nothing #hide
+
+# However, for generic structuring element, the half-size (or, radius) should be at most `1`
+# for each dimension. This is because the algorithm is a growing process until steability,
```
Comment (ThomasRetornaz) [inline docs/operators/operators/mreconstruct.jl:106]: typo steability -> stability

**Comment by ThomasRetornaz** [comment]
> Without changing the algorithm, optimizations that I can think of is
> 
>     * pad the marker image with sentinel values so that we can skip bounds check and save some computation (This boosts about 15% of the performance). Given that the loop order must be fixed, SIMD and other loop reorder tricks (e.g., `@turbo`) can hardly be applied here.
> 
>     * short-circuit the loop if we know in prior that it's already the maximal/minimal value. 
>     * specialized implementation for particular SE shapes. For instance, for 2D diamond shape SE, we can manually inline the `upper_se` and `lower_se` to avoid the array `getindex`. But I'm not sure how much performance we can get.

Yes i don't think we could "trivially" optimize the algorithm of vincent :( 
One thing to try is to do few steps <10 of naive algorithms (very fast for SE 2D diamond and other incoming SE) but its not really satisfaying and complexify dispatch  

in any case thank you for this PR, very impressive doc and test coverage
Regards

**Pull request merged**